### ISSUE TITLE
[Proposal] LocalizationResourceManager

### ISSUE THREAD

**Issue by brminnick** [CommunityToolkit/Maui#93]
# LocalizationResourceManager

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
  * [ ] iOS Support
  * [ ] Android Support
  * [ ] macOS Support
  * [ ] Windows Support
* [ ] Unit Tests: Not Started
* [ ] Sample: Not Started
* [ ] Documentation: Not Started

## Summary
[summary]: #summary

The LocalizationResourceManager class is a helper class that enables users to respond to culture changes at runtime. This class is typically used by the TranslateExtension class and LocalizedString.

## Detailed Design
[design]: #detailed-design

### LocalizationResourceManager.shared.cs

```cs
public class LocalizationResourceManager : ObservableObject
{
  public static LocalizationResourceManager Current;

  private LocalizationResourceManager();

  public void Init(ResourceManager resource);
  public void Init(ResourceManager resource, CultureInfo initialCulture);
  public string GetValue(string text);

  public string this[string text] => GetValue(text);

  public CultureInfo CurrentCulture { get; set; }
}
```

## Usage Syntax
[usage]: #usage-syntax

### XAML Usage

N/A

### C# Usage

```cs
LocalizationResourceManager.Current.PropertyChanged += (_, _) => AppResources.Culture = LocalizationResourceManager.Current.CurrentCulture;
LocalizationResourceManager.Current.Init(AppResources.ResourceManager);

LocalizationResourceManager.Current.CurrentCulture = newCulture;
```

## Drawbacks
[drawbacks]: #drawbacks

This feature seems to duplicate the built-in localization tools with RESX.

It also implements `ObservableObject` which will not be included in `CommunityToolkit.Maui` because it is already implemented in `CommunityToolkit.Mvvm`.

## Alternatives
[alternatives]: #alternatives

**RESX**
https://docs.microsoft.com/en-us/xamarin/xamarin-forms/app-fundamentals/localization/text?pivots=macos?WT.mc_id=mobile-0000-bramin

```cs
CultureInfo.DefaultThreadCurrentCulture = "en-us";
CultureInfo.DefaultThreadCurrentUICulture = "en-us";
```

**Comment by brminnick** [comment]
Closing for the following reasons: 
- Duplication of RESX 
- Required use of `ObservableObject`
  - `ObservableObject` is not  implemented in `CommunityToolkit.Maui`
  - `ObservableObject` is implemented in `CommunityToolkit.MVVM`

**Comment by brminnick** [comment]
Reopening based on our discussion during the June Standup:  https://www.youtube.com/watch?v=t3g_NrQfE8g

**Comment by VladislavAntonyuk** [comment]
Why I suggest to keep it in MCT?
1. It works great with TranslateExtension, which depends on MAUI. So if we decide to add TranslateExtension we will need to depend on Community Toolkit dotnet.
2. Based on the first point. Splitting Localization classes on different packages maybe not a great customer experience 
3. Minor but it will be easier for XCT users migrate to MCT
4. As for me localization resources are mostly related to UI (like button text or user error messages - not logs), so keep it somewhere in Core library may not give a lot of benefit. User may store Resx files wherever he wants. LocalizationResourceManager only needs ResourceManager reference.

**Comment by bijington** [comment]
I agree that we should probably offer all or nothing rather than take on a dependency. I am trying to think on whether there could be a compromise but nothing is coming to mind

**Comment by VladislavAntonyuk** [comment]
I wrote an article demonstrating the benefit of releasing LocalizationResourceManager and TranslateExtension together: https://vladislavantonyuk.azurewebsites.net/articles/Localize-.NET-MAUI-application

**Comment by pictos** [comment]
For me, we should implement this in a more `Core` lib. in other to add value across other .NET UI frameworks (and maybe in non UI frameworks as well). I'll express that by commenting on the points that @VladislavAntonyuk mentioned.

So from 1 and 2
> 1. It works great with TranslateExtension, which depends on MAUI. So if we decide to add TranslateExtension we will need to depend on Community Toolkit dotnet.
> 2. Based on the first point. Splitting Localization classes on different packages maybe not a great customer experience

Now that CommunityToolkits are all in the same org, we should start thinking about ways to add features to all of them. Like the `CommunityToolkit.Maui.Core` will be used by other Toolkits that you will target other MAUI flavors, we can have a dependency on the `CommunityToolkit.Common` if that makes sense.
We can ship the `CommunityToolkit.Common` with our NuGet, so devs don't need to install it by themself.

> 3. Minor but it will be easier for XCT users migrate to MCT

Can't see why will be hard to migrate from XCT to MCT if this feature is on `CommunityToolkit.Common`. Repeating my self, we can ship that package with our NuGet.


> 4. As for me localization resources are mostly related to UI (like button text or user error messages - not logs), so keep it somewhere in Core library may not give a lot of benefit. User may store Resx files wherever he wants. LocalizationResourceManager only needs ResourceManager reference.

You can use resx files in ConsoleApps (maybe you're developing a `dotnet tool` and want to support multiple languages), and also can be used in other .NET areas, that include UI, like ASP.NET, WPF, UWP, etc. I mean, MAUI isn't the only UI framework for .NET so having this feature in a `Core` library will help other toolkits, applications and libs to create and manage their localization resources like we want to do here

**Comment by brminnick** [comment]
Because this Proposal is not specific to .NET MAUI, my vote is to implement this feature in the CommunityToolkit/dotnet repo so that additional devs can benefit from it. 

Hey @sergio0694! Do agree that this feature is best scoped for `CommunityToolkit.Common`?

**Comment by Sergio0694** [comment]
Leaving an update here as @brminnick asked - right now the 8.0 release is feature complete and I'm just working on stabilizing stuff and fixing last minute bugs. I'm happy to look into this once that is out though, for sure 🙂
